File: Ruby/vista_textual.rb
#encoding:utf-8
require_relative "operaciones_juego.rb"
require 'io/console'
require_relative "salidas_carcel.rb"
require_relative "respuestas.rb"
require_relative "diario.rb"

module Civitas

  class Vista_textual

    def mostrar_estado(estado)
      puts estado
    end

    
    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end



    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l
        index += 1
      }

      opcion = lee_entero(lista.length,
        "\n"+tab+"Elige una opción: ",
        tab+"Valor erróneo")
      return opcion
    end

    def salirCarcel  #PENDIENTE DE DEPURACIÓN!!!
      lista_opciones=["Pagando","Tirando"]
      iOpcion=menu("Elige como quiere salir de la cárcel",lista_opciones)
      return SalidasCarcel::Lista_SalidasCarcel[iOpcion]
    end
    
    def comprar #PENDIENTE DE DEPURACIÓN!!!
      lista_opciones=["No","Sí"]
      iOpcion=menu("¿Quieres comprar la propiedad?", lista_opciones)
      return Respuestas::Lista_Respuestas[iOpcion]
    end

    def gestionar #PENDIENTE DE DEPURACIÓN!!!
      lista_opciones=["Vender", "Hipotecar", "Cancelar hipoteca", "Construir casa", "Construir hotel", "Terminar"]
      @iGestion = menu("Elige el número de gestión inmobiliaria", lista_opciones)
      nombres = @juegoModel.jugadores[@juegoModel.indideJugadorActual].getNombrePropiedades
      @iPropiedad = menu("Elige la propiedad en la que realizar la acción", nombres)
    end
    endLista_Respuestas

    def getGestion
      return @iGestion
    end

    def getPropiedad
      return @iPropiedad
    end

    def mostrarSiguienteOperacion(operacion)
      puts operacion
    end

    def mostrarEventos #PENDIENTE DE DEPURACION!!!
      while(Diario.instance.eventos_pendientes)
        puts Diario.instance.leer_evento
      end
    end

    def setCivitasJuego(civitas)
      @juegoModel=civitas
      actualizarVista
    end

    def actualizarVista #PENDIENTE DE DEPURACIÓN
      indiceActual = @juegoModel.indideJugadorActual
      puts @juegoModel.jugadores[indiceActual].to_s
      puts @juegoModel.tablero.casillas[@juegoModel.jugadores[indiceActual].numCasillaActual].to_s
      nombres = Array.new
      nombres = @juegoModel.jugadores[indiceActual].getNombrePropiedades
      for i in(0..nombres.size-1)
        puts nombres[i]
      end
    end

    
  end

end

File: juegoTexto/VistaTextual.java
package juegoTexto;

import civitas.CivitasJuego;
import civitas.Diario;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import civitas.Casilla;
import civitas.Jugador;
import civitas.TituloPropiedad;
import GUI.Respuestas;

public class VistaTextual {
  
  CivitasJuego juegoModel; 
  int iGestion=-1;
  int iPropiedad=-1;
  private static String separador = "=====================";
  
  private Scanner in;
  
  public VistaTextual () {
    in = new Scanner (System.in);
  }
  
  void mostrarEstado(String estado) {
    System.out.println (estado);
  }
              
  public void pausa() {
    System.out.print ("Pulsa una tecla");
    in.nextLine();
  }

  int leeEntero (int max, String msg1, String msg2) {
    Boolean ok;
    String cadena;
    int numero = -1;
    do {
      System.out.print (msg1);
      cadena = in.nextLine();
      try {  
        numero = Integer.parseInt(cadena);
        ok = true;
      } catch (NumberFormatException e) { // No se ha introducido un entero
        System.out.println (msg2);
        ok = false;  
      }
      if (ok && (numero < 0 || numero >= max)) {
        System.out.println (msg2);
        ok = false;
      }
    } while (!ok);

    return numero;
  }

  int menu (String titulo, ArrayList<String> lista) {
    String tab = "  ";
    int opcion;
    System.out.println (titulo);
    for (int i = 0; i < lista.size(); i++) {
      System.out.println (tab+i+"-"+lista.get(i));
    }

    opcion = leeEntero(lista.size(),
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo");
    return opcion;
  }

  public SalidasCarcel salirCarcel() {
    int opcion = menu ("Elige la forma para intentar salir de la carcel",
      new ArrayList<> (Arrays.asList("Pagando","Tirando el dado")));
    return (SalidasCarcel.values()[opcion]);
  }

  public Respuestas comprar() {
    int opcion = this.menu("¿Quieres comprar la propiedad?", new ArrayList<> (Arrays.asList("No","Sí")));
    return (Respuestas.values()[opcion]);
  }

  public void gestionar() {
  
      this.iGestion = menu("Elige el número de gestión inmobiliaria", new ArrayList<> (Arrays.asList("Vender","Hipotecar", "Cancelar hipoteca", "Construir casa", "Construir hotel", "Terminar")));
      ArrayList<String> nombres;
      nombres = new ArrayList<>();
      nombres = this.juegoModel.getJugadorActual().getNombrePropiedades();
      this.iPropiedad = menu("Elige la propiedad en la que realizar la acción.", nombres);
  }
  
  public int getGestion(){
      
      return this.iGestion;
  }
  
  public int getPropiedad(){
      return this.iPropiedad;
  }
    

  public void mostrarSiguienteOperacion(OperacionesJuego operacion) {
  
       System.out.println(operacion);
  }


  public void mostrarEventos() {
  
      while(Diario.getInstance().eventosPendientes()){
          System.out.println(Diario.getInstance().leerEvento());
      }
  }
  
  public void setCivitasJuego(CivitasJuego civitas){ 
        juegoModel=civitas;
        this.actualizarVista();
    }
  
  public void actualizarVista(){
      System.out.println(this.juegoModel.getJugadorActual().toString());
      System.out.println(this.juegoModel.getCasillaActual().toString());
      ArrayList<String> nombres;
      nombres = new ArrayList<>();
      nombres = this.juegoModel.getJugadorActual().getNombrePropiedades();
      for(int i=0; i<nombres.size(); i++){
          System.out.println(nombres.get(i)+" "+this.juegoModel.getJugadorActual().getPropiedades().get(i).getNumCasas());
          System.out.println("\n");
      }
  } 
}
